packages/seacas/libraries/ioss/src/exodus/Ioex_BaseDatabaseIO.C

```diff
@@ -95,6 +95,76 @@ namespace {
   void generate_block_truth_table(Ioex::VariableNameMap &variables, Ioss::IntVector &truth_table,
                                   std::vector<T *> &blocks, char field_suffix_separator);
 
+  void insert_sort_and_unique(const std::vector<std::string>& src, std::vector<std::string>& dest);
+
+  class AssemblyTreeFilter
+  {
+  public:
+    AssemblyTreeFilter(Ioss::Region* region, const Ioss::EntityType filterType,
+                       const std::vector<ex_assembly>& assemblies)
+    : m_region(region)
+    , m_type(filterType)
+    , m_assemblies(assemblies)
+    , m_visitedAssemblies(assemblies.size(), false) {}
+
+    void update_list_from_assembly_tree(size_t assemblyIndex, std::vector<std::string>& list)
+    {
+      // Walk the tree without cyclic dependency
+      if (assemblyIndex < m_assemblies.size()) {
+        if (m_visitedAssemblies[assemblyIndex] == false) {
+          m_visitedAssemblies[assemblyIndex] = true;
+
+          const auto &assembly = m_assemblies[assemblyIndex];
+          const Ioss::EntityType assemblyType = Ioex::map_exodus_type(assembly.type);
+          if(m_type == assemblyType) {
+            for (int j = 0; j < assembly.entity_count; j++) {
+              Ioss::GroupingEntity *ge = m_region->get_entity(assembly.entity_list[j], m_type);
+              if(nullptr != ge) {
+                list.push_back(ge->name());
+              }
+            }
+          }
+
+          if(Ioss::ASSEMBLY == assemblyType) {
+            for (int i = 0; i < assembly.entity_count; i++) {
+              // Find the sub assembly with the same id
+              int64_t subAssemblyId = assembly.entity_list[i];
+              bool found = false;
+              for(size_t j=0; j<m_assemblies.size(); j++) {
+                if(m_assemblies[j].id == subAssemblyId) {
+                  found = true;
+                  update_list_from_assembly_tree(j, list);
+                }
+              }
+              assert(found);
+            }
+          }
+        }
+      }
+    }
+
+    void update_assembly_filter_list(std::vector<std::string>& assemblyFilterList)
+    {
+      for (size_t i=0; i<m_assemblies.size(); ++i) {
+        if(m_visitedAssemblies[i]) {
+          assemblyFilterList.push_back(m_assemblies[i].name);
+        }
+      }
+
+      std::sort(assemblyFilterList.begin(), assemblyFilterList.end(), std::less<std::string>());
+      auto endIter = std::unique(assemblyFilterList.begin(), assemblyFilterList.end());
+      assemblyFilterList.resize(endIter - assemblyFilterList.begin());
+    }
+
+  private:
+    AssemblyTreeFilter() = delete;
+    AssemblyTreeFilter(const AssemblyTreeFilter&) = delete;
+
+    Ioss::Region* m_region;
```

Comment:
Initialize to nullptr

Reply:
I can do that also. As I pointed out earlier, making the object non-default and non-copy constructible means that the only way to create it is through the specified constructor and that does the initialization of the class members. My eclipse IDE flags data members that are not initialized properly

Reply:
Understood.  I like to still initialize them just in case someone changes them to be default or copy constructible or an operator= gets generated or added.